### PR TITLE
Garnettanner/stm32 example

### PR DESCRIPTION
**Pull request by Gtanner00**
Updated STM32F411e_Example with corrections to the delay code as discussed with Enoch.

Added an Asciidoc help for setting up the build environment on linux.

### CODE REVIEW

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 1]
```diff
@@ -0,0 +1,81 @@
+
+
+= *Setting up and Using an STM32 Build Environment on Linux*
+A guide for setting up a linux system to build projects for STM32
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:4]: Asciidoc should should have a space after headers

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 2]
```diff
@@ -0,0 +1,81 @@
+
+
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:2]: These empty lines should be removed

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+= *Setting up and Using an STM32 Build Environment on Linux*
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:3]: Headers don't need to be bolded.

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+A guide for setting up a linux system to build projects for STM32
+
+
+This guide assumes you are starting with a Windows 64 bit system and intend to use a virtual machine in Oracle Virtualbox. I installed Linux Mint 18.3 If you already have a linux system set up, are using a Mac, or want to set it up a diffeent way figure out what parts of this guide you have already completed yourself and rejoin at a later step.
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:7]: Good asciidoc practice is having one sentence per line. And in general, documentation should be written in third person without personal anectodes. For example this excerpt would be better written as,


```asciidoc
This guide assumes the base operating system is 64 bit Windows.
The steps include setting up a virtual machine in Oracle VirtualBox, but if a Linux system is already set up in any fashion, this step can be skipped.
The Linux system that this guide has been written with is Linux Mint 18.3.
```

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+== Install Virtualbox
+Download https://www.virtualbox.org/wiki/Downloads[Oracle Virtual Box]
+The latest version at writting was 5.2.4. Also download the VirtualBox Extension Pack from the same page.
+Install both.
+
+==== Setup a Virtual Machine
+Create a new virtual machine in Virtualbox.
+
+image::https://github.com/StudentOrganisationForAerospaceResearch/DevelopmentResources/blob/garnettanner/stm32-example/STM32Resources/BuildEnvironment_Setup/images/1.PNG[]
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:17]: For linking images, they can be effective with relative paths, and by doing so, it would also help the link to be more robust.

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+This guide assumes you are starting with a Windows 64 bit system and intend to use a virtual machine in Oracle Virtualbox. I installed Linux Mint 18.3 If you already have a linux system set up, are using a Mac, or want to set it up a diffeent way figure out what parts of this guide you have already completed yourself and rejoin at a later step.
+
+== Install Virtualbox
+Download https://www.virtualbox.org/wiki/Downloads[Oracle Virtual Box]
+The latest version at writting was 5.2.4. Also download the VirtualBox Extension Pack from the same page.
+Install both.
+
+==== Setup a Virtual Machine
+Create a new virtual machine in Virtualbox.
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:15]: This step to me sounds too general, it can be more specific I think. 

```asciidoc
The first step is to create a new virtual machine in Virtualbox, this can be done by hitting the _New_ icon.
```

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+image::https://github.com/StudentOrganisationForAerospaceResearch/DevelopmentResources/blob/garnettanner/stm32-example/STM32Resources/BuildEnvironment_Setup/images/1.PNG[]
+
+There isn't an option or Linux Mint OS so choose Ubuntu.
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:19]: I would provide the information that Linux Mint is based off of Ubuntu, so the selected OS can be Ubuntu.

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+You'll need 20-30+GB of hard disk space to hold the OS, tools and your projects.
+Run through the creation of your VM and hard disk configuration
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:21]: A list of recommended settings or saying that all the defaults are fine I think would be helpful.

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+When Virtualbox asks to select a start up disk, navigate to the ISO which will boot into the installer.
+Click the disk icon to start the installation and run through it.
+
+image::https://github.com/StudentOrganisationForAerospaceResearch/DevelopmentResources/blob/garnettanner/stm32-example/STM32Resources/BuildEnvironment_Setup/images/2.PNG[]
+
+image::https://github.com/StudentOrganisationForAerospaceResearch/DevelopmentResources/blob/garnettanner/stm32-example/STM32Resources/BuildEnvironment_Setup/images/3.PNG[]
+
+==== Configure Virtual Machine
+You'll need to enable a fe
+w features in Virtualbox to allow everything to work.
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:36]: A cut off sentence here haha

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+
+This guide assumes you are starting with a Windows 64 bit system and intend to use a virtual machine in Oracle Virtualbox. I installed Linux Mint 18.3 If you already have a linux system set up, are using a Mac, or want to set it up a diffeent way figure out what parts of this guide you have already completed yourself and rejoin at a later step.
+
+== Install Virtualbox
+Download https://www.virtualbox.org/wiki/Downloads[Oracle Virtual Box]
+The latest version at writting was 5.2.4. Also download the VirtualBox Extension Pack from the same page.
+Install both.
+
+==== Setup a Virtual Machine
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:14]: This should be a level 3 or 2 header.

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+Create a new virtual machine in Virtualbox.
+
+image::https://github.com/StudentOrganisationForAerospaceResearch/DevelopmentResources/blob/garnettanner/stm32-example/STM32Resources/BuildEnvironment_Setup/images/1.PNG[]
+
+There isn't an option or Linux Mint OS so choose Ubuntu.
+You'll need 20-30+GB of hard disk space to hold the OS, tools and your projects.
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:20]: I think the best answer to give here would be a specific minimum amount. 

```asciidoc
The recommended minimum disk space is 20GB, but it is recommended to have at least 40GB.
```

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+----
+
+==== STLINK, CubeMXMakefile
+Clone the Development Resources repo on the SOAR Git. Open the terminal and navigate
+to where you want to clone to.
+----
+git clone https://github.com/StudentOrganisationForAerospaceResearch/DevelopmentResources.git
+----
+Find install.sh in DevelopmentREsources/STM32Resources.
+Change the permissions for the file to allow executing as a program. See the properties menu.
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:56]: Including the command would probably be helpful :) `chmod +x install.sh`

Or even better, we push up a fix to the file to already have executable permissions.

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+Shut down your VM. Go into settings and enable 3D acceleration for the display (if available) and the USB controller. When you plug in a USB device, you can capture it in the VM by clicking the USB icon in the bottom right of the window and selecting your device.
+
+==== Install Git and other Packages
+==== Git
+----
+ sudo apt-get install git-all
+----
+
+==== STLINK, CubeMXMakefile
+Clone the Development Resources repo on the SOAR Git. Open the terminal and navigate
```
Comment (ExtraGravity) [inline STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:50]: `on the SOAR Git` is kinda of awkward sounding. I think `...from the SOAR Github Organization` would be better.

File: STM32Resources/install.sh [hunk 14]
```diff
@@ -6,7 +6,19 @@ if [[ $UID != 0 ]]; then
     exit 1
 fi
 
-apt-get -y install build-essential git libsane:i386 ia32-libs-multiarch autoconf libusb-1.0-0-dev lib32ncurses5 libncurses5:i386 software-properties-common pkg-config cmake
+apt-get update
```
Comment (ExtraGravity) [inline STM32Resources/install.sh:9]: Excellent fix, can you put this into a separate PR so we can merge it first?

**Review by ExtraGravity — CHANGES_REQUESTED** [review]
Thanks for writing this. Really good for a first time writing asciidoc. For organization purposes, I would actually put all guides under the Documentation folder, a README here could link to the proper READMEs. 

Along that line as well, I would put instructions for creating a linux virtual machine as a separate instruction document since it is not specific to a dev environment for the STM32.

Also, I would put the PR for the example project and the PR for the guide separately. This way we can merge one before the other. Usually if a PR can be separated into multiple PRs (i.e. when there are non dependent changes), they should be in separate PRs. 

Also there are a fair amount of files from the example project that should be ignored. For example the entire `build` folder should be ignored. 

I've said a lot already so you can make these changes first, then I will probably do some more nitpicking :) 
Really good job though Garnet, you're impressive. 

TL;DR Split up the PRs first, and make the changes. I think I'll actually do the PR on the example project so you can leave out that one.

 I reviewed my guide for creating a pull request and it is sound, so if you run into issues again, let's work through it so I know where in the documentation would be slipping people up.

File: STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc (hunk 5 above)
Comment (Gtanner00) [reply STM32Resources/BuildEnvironment_Setup/BuildEnvironment_Setup.adoc:17]: I could not get images to show up correctly using relative paths. I think it's a github limitation.

File: STM32Resources/install.sh (hunk 14 above)
Comment (Gtanner00) [reply STM32Resources/install.sh:9]: What does this do?

Comment (ExtraGravity) [reply STM32Resources/install.sh:9]: This command updates the source locations for all the packages, so updates where to download them and stuff, kind of. So before running apt-get install commands it's usually a good idea to include it or users can get weird errors, especially on fresh installs.
